add time clock to runtime

Joeperdefloep opened the pull request:
<!-- Feel free to remove check-list items aren't relevant to your change -->
added placeholder, should work now
 - [ ] Closes #155 
 - [ ] Tests added
 - [ ] Passes `black . && flake8`
 - [ ] Fully documented, including `whats-new.rst` for all changes and `api.rst` for new API

benbovy commented:
Thanks @Joeperdefloep! However, I don't think `MAIN_CLOCK` will add anything really useful in this PR.

The "placeholder" that I suggest in https://github.com/benbovy/xarray-simlab/issues/155#issuecomment-775507738 is for addressing the case where we want to use the main clock as an explicit dimension when declaring xarray-simlab variables in process classes. It is a different (but related) problem than this PR. We would need something a bit more advanced than a string, i.e., a custom type that xarray-simlab understands so that it can "replace" it with the actual dimension label corresponding to the main clock (later set by the user) when creating xarray Dataset objects. Although for an entirely different purpose, an example of such type (sentinel class, singleton) is [attr.NOTHING](https://github.com/python-attrs/attrs/blob/c2712fd102ddfb7df6df333f421c3b21c18e4b37/src/attr/_make.py#L55-L84).

Updating runtime context entries with the main clock values (this PR) should be addressed afterwards IMO. Runtime context entries are already hard-coded, so we can just use the `_main_clock` and `main_clock` strings.

benbovy commented:
Hmm I'm not sure to understand what you have implemented you last commit.

Just to be sure, what we'd like to do is something like the small example below or do you have something else in mind?

```python
@xs.process
class Foo:
    var = xs.variable(dims=xs.MAIN_CLOCK, intent='out')

    @xs.runtime(args='main_clock_array')
    def initialize(self, main_clock):
        self.var = main_clock * 2


m = xs.Model({'foo': Foo})

ds_in = xs.create_setup(
    model=m,
    clocks={'time': range(5)},
    output_vars={'foo__var': None}
)

ds_out = ds_in.xsimlab.run(model=m)

# should print a DataArray with dimension 'time'
# and values [0, 2, 4, 6, 8]
print(ds_out.foo__var)
```

For this example, we need to implement new functionality in xarray-simlab in order to support the two lines below:

1. `var = xs.variable(dims=xs.MAIN_CLOCK, intent='out')`
2. `@xs.runtime(args='main_clock_array')`

This PR implements 2, and I suggest implementing 1 in a separate PR (Ideally, we should address 1 first) 

What I have in mind for 1 is just a very basic placeholder `xs.MAIN_CLOCK` that does nothing else than tell xarray-simlab to get the main clock dimension label and use it to create the `foo__var` output variable in the simulation store. `xs.MAIN_CLOCK` thus don't hold any actual dimension label and would never be used in any xarray object.

Joeperdefloep commented:
>Just to be sure, what we'd like to do is something like in the small example below or do you have something else in mind?

Yes, that was exactly what I wanted to implement! Well, actually a bit more:
I also implemented 2, to be able to debug with a small example:
```python

@xs.process
class dBiomass:
    """calculates biomass change based on light intensity
    """
    maxrad = xs.variable(dims=xs.MAIN_CLOCK,intent='out')
    biomass = xs.variable(intent='out')
    
    @xs.runtime(args=['main_clock_values','main_clock_array'])
    def initialize(self,clock_values,clock_array):
        print(clock_values)
        print(clock_array[xs.MAIN_CLOCK])
        self.biomass=0
        self.maxrad = main_clock
    
    @xs.runtime(args='step_delta')
    def run_step(self,dt):
        self._d_biomass = dt*self.maxrad
        
    def finalize_step(self):
        self.biomass += self._d_biomass

biomass_model_raw = xs.Model({'dBiomass':dBiomass})

ds_in = xs.create_setup(
    model=biomass_model_raw,
    clocks={
        'day':range(100)
    },
    master_clock='day',
    input_vars={},
    output_vars={'dBiomass__biomass':'day'}
)
```
maybe end-to-end testing is not the most efficient way of adding a new feature, how do you normally tackle such things, or how would I go about debugging 1. without first implementing 2? Writing tests maybe? :speak_no_evil: Serious question though...

this implements the following lines:
1. `maxrad = xs.variable(dims=xs.MAIN_CLOCK,intent='out')` This works correctly, since `xs.MAIN_CLOCK` is updated to the `master_clock_dim` in the `update_vars` step. You could `print(xs.MAIN_CLOCK)` in any process step and get `day`.
2. `@xs.runtime(args='main_clock_values')`: access to main clock as a raw array  and subsequent processing. Here are some problems: Since normally only time-dependent variables are set as pre-processing, we need to add logic that expands the vars along the time dimension if a time dimension is set in the var. This could lead to problems if:
    1.  multiple options are given, such as `dims=[(x,y),(x,xs.MAIN_CLOCK)]`
    2. the dimension of `y` is as long as `xs.MAIN_CLOCK`
Again, we could check for this, in the expand-along-time-dimension logic that should go in the initialize step. I thought this could maybe be avoided, if the user passes a `xr.DataArray`, with explicit dimensions, which brings us to the next:
3. @xs.runtime(args='main_clock_array')`: access to main clock as a DataArray. This also implies use of `clock.array[xs.MAIN_CLOCK]`. This may only be necessary when #141 is fully addressed, but it is the reason for the singleton class in favour of a simple string placeholder that is replaced by the actual string. This will not be updated accordingly between definition and problem run. Indeed, maybe this should be a separate pull request, (but maybe it is nice to have it for when #141 is to be implemented?)

There is another possible

benbovy commented:
> how do you normally tackle such things, or how would I go about debugging 1. without first implementing 2?

Isn't this PR and #155 initially about implementing 2? You could test 2 without having 1 implemented if that's you mean, e.g.,

```python
@xs.process
class Foo:
    var = xs.variable(dims=(), intent='out')

    @xs.runtime(args='main_clock_array')
    def initialize(self, main_clock):
        self.var = np.sum(main_clock)


m = xs.Model({'foo': Foo})

ds_in = xs.create_setup(
    model=m,
    clocks={'time': range(5)},
    output_vars={'foo__var': None}
)

ds_out = ds_in.xsimlab.run(model=m)

assert ds_out.foo__var == 10
```

> This works correctly, since xs.MAIN_CLOCK is updated to the master_clock_dim in the update_vars step.

One major drawback is that this approach is not thread safe.

> Since normally only time-dependent variables are set as pre-processing, we need to add logic that expands the vars along the time dimension if a time dimension is set in the var. 

Not sure I understand, but I guess you mean that we should take care of the case where users provide time-varying values for a variable like `xs.variable(dims=xs.MAIN_CLOCK, intent='in')`? We probably should. I suggest that `xs.variable(dims=xs.MAIN_CLOCK, intent='in')` implicitly implies `xs.variable(dims=xs.MAIN_CLOCK, intent='in', static=True)` (or just check that `static=True` if we want to avoid being implicit), so that no time-varying input value is allowed. (Note: by time-varying value I mean provide an input array with one additional dimension that is not defined in the variable `dims`).

> multiple options are given, such as dims=[(x,y),(x,xs.MAIN_CLOCK)]

This is not possible in xarray-simlab, each option must have a different number of dimensions.

> @xs.runtime(args='main_clock_array'): access to main clock as a DataArray. This also implies use of clock.array[xs.MAIN_CLOCK]`

It would certainly be useful to access the main clock as a DataArray, but why `clock.array[xs.MAIN_CLOCK]`? Why not just setting `main_clock_array=dataset.xsimlab.main_clock_coord` [here](https://github.com/benbovy/xarray-simlab/blob/c56d465f1893e6c1ca77f213902dad69de043f69/xsimlab/drivers.py#L324)?

> There is another possible problem when a process dimension happens to have the same name as the main clock dimension e.g. xs.variable(dims=(xs.MAIN_CLOCK,'day'))

We should raise an error in this case. It either to the model user to choose another label for the main clock, or the model developer to use another, potentially less conflicting label in process classes.

Joeperdefloep commented:
> Isn't this PR and #155 initially about implementing 2? You could test 2 without having 1 implemented if that's what you mean, e.g.,

Well, it is about making the simple example work. The question was how to test 1 without 2. But I think the following example does that (testing 1 without access to `main_clock_values`):

```python
time_array = np.array([0,1,2,3])

@xs.process
class Foo:
    #intent cannot be 'in' if we want to set it in initialize
    bar = xs.variable(dims=xs.MAIN_CLOCK,intent='in',static=True)
    baz = xs.variable(intent='out')
    
    def initialize(self):
        self.bar = time_array*2
        self.baz=0
    
    @xs.runtime(args='step_delta')
    def run_step(self,dt):
        #now self.bar is actually [0,2,4,6]
        #should be 0-6 in different steps
        self.baz += dt*self.bar

m = xs.Model({'foo':Foo})


ds_in = xs.create_setup(
    model=m,
    clocks={
        'time':time_array
    },
    input_vars={
        'foo__bar': 5,
    },
    output_vars={
        #add output var to get error: dims don't match
#         'Biomass__biomass':'time'
    }
)

ds_in.xsimlab.run(m)
```

> > This works correctly, since xs.MAIN_CLOCK is updated to the master_clock_dim in the update_vars step.
> 
> One major drawback is that this approach is not thread safe.

ok yes, if a user was to create multiple models with multiple with different main_clock dimensions, and processes that refer to `xs.MAIN_CLOCK`, then these different threads would be setting `xs.MAIN_CLOCK` to different values. However, this implementation _is_ safe against the normal multiprocessing, since it is set in preprocessing.
An alternative would be to add MAIN_CLOCK to each process, so that:
1. `xs.MAIN_CLOCK` is instantiated when a model is created. All earlier references to `xs.MAIN_CLOCK` now point to that class instance. this sounds like becoming really complex to me: How do we know which references are from our model and which are not? Should we put the MAIN_CLOCK in the process class?
2. it is still updated in `update_clocks` to match the `main_clock_dim`. This is a preprocessing step, and therefore safe for batch model runs

> > Since normally only time-dependent variables are set as pre-processing, we need to add logic that expands the vars along the time dimension if a time dimension is set in the var.
> 
> Not sure I understand, but I guess you mean that we should take care of the case where users provide time-varying values for a variable like `xs.variable(dims=xs.MAIN_CLOCK, intent='in')`? We probably should. I suggest that `xs.variable(dims=xs.MAIN_CLOCK, intent='in')` implicitly implies `xs.variable(dims=xs.MAIN_CLOCK, intent='in', static=True)` (or just check that `static=True` if we want to avoid being implicit), so that no time-varying input value is allowed. (Note: by time-varying value I mean provide an input array with one additional dimension that is not defined in the variable `dims`).

actually, setting `intent='in'` variables is not allowed in the initialize step (or any step). I think we should keep it that way, and solve #155 for 'out' or 'inout' variables, since we calculate and update them (in the initialize step). In fact, a user cannot access time-varying inputs in the `initialize` step of a model(footnote).

What I meant here is that in the above example, in `initialize`, we set the value as an array, whereas in `run_step`, the values should be accessed as individual values. Therefore, in (after) the initialize step, variables that have `xs.MAIN_CLOCK` in their dims, should expand it over that axes, so that individual values are passed in `run_step`.

> It would certainly be useful to access the main clock as a DataArray, but why `clock.array[xs.MAIN_CLOCK]`? Why not just setting `main_clock_array=dataset.xsimlab.main_clock_coord` [here](https://github.com/benbovy/xarray-simlab/blob/c56d465f1893e6c1ca77f213902dad69de043f69/xsimlab/drivers.py#L324)?

I did that there, and changed the dimesnions from `_main_clock_array` to `dataset.xsimlab.master_clock_dim`. The comment was from a user-perspective. For example if the user has clock in a datetime and wants to be able to `clock_array[xs.MAIN_CLOCK + '.dayofyear']` or use any operation such as `.loc` at runtime.
 
> > There is another possible problem when a process dimension happens to have the same name as the main clock dimension e.g. xs.variable(dims=(xs.MAIN_CLOCK,'day'))
> 
> We should raise an error in this case. It either to the model user to choose another label for the main clock, or the model developer to use another, potentially less conflicting label in process classes.

footnote:
Keyerror when accessing time-varying data in initialize step:
```python
@xs.process
class Foo:
    bar = xs.variable(intent='in')
    baz = xs.variable(intent='out')
    
    def initialize(self):
        #this gives a keyerror on time-varying data
        #could be solved with #155
        self.baz=self.bar
    
    @xs.runtime(args='step_delta')
    def run_step(self,dt):
        self.baz += dt*self.bar

m = xs.Model({'foo':Foo})


ds_in = xs.create_setup(
    model=m,
    clocks={
        'time':time_array
    },
    input_vars={
        'foo__bar': 5,
    },
    output_vars={
        'foo__baz':'time'
    }
)

# ds_in = ds_in.xsimlab.update_vars(model=m,input_vars={'foo':{'bar':ds_in.time*5}})
ds_in.xsimlab.run(m).foo__baz.plot()
```

benbovy commented:
> The comment was from a user-perspective. For example if the user has clock in a datetime and wants to be able to clock_array[xs.MAIN_CLOCK + '.dayofyear'] or use any operation such as .loc at runtime.

If `clock_array` is a DataArray with one dimension that corresponds to the main clock, we could just get it from `clock_array.dims`, no?

> What I meant here is that in the above example, in initialize, we set the value as an array, whereas in run_step, the values should be accessed as individual values. Therefore, in (after) the initialize step, variables that have xs.MAIN_CLOCK in their dims, should expand it over that axes, so that individual values are passed in run_step.

Ok I see. Allow declaring input variables with a `xs.MAIN_CLOCK` dimension would potentially bring a lot of issues. I really need to think more about this. As you say it's probably wiser to allow it for output variables only.

For the case of output variables set in `initialize`, if we want to access the array element value for the current step in `run_step`,  we could just reuse `@xs.runtime(args='step')`.

Joeperdefloep commented:
Ok, so I properly implemented the singleton class (thanks for pointing out the right places! I was really at a loss there, as to how to implement it, it was indeed still very ugly).

> If `clock_array` is a DataArray with one dimension that corresponds to the main clock, we could just get it from `clock_array.dims`, no?

hmm, maybe you're right. I added a link to the dim in `drivers`, can remove it again if you want.

> > What I meant here is that in the above example, in initialize, we set the value as an array, whereas in run_step, the values should be accessed as individual values. Therefore, in (after) the initialize step, variables that have xs.MAIN_CLOCK in their dims, should expand it over that axes, so that individual values are passed in run_step.
> 
> Ok I see. Allow declaring input variables with a `xs.MAIN_CLOCK` dimension would potentially bring a lot of issues. I really need to think more about this. As you say it's probably wiser to allow it for output variables only.

I added a check/ValueError in `_maybe_transpose`, that is always called to check the dims. to warn the user.

> For the case of output variables set in `initialize`, if we want to access the array element value for the current step in `run_step`, we could just reuse `@xs.runtime(args='step')`.

Joeperdefloep commented:
Now, when we push, some (2) `master` need to be changed to `main`

Joeperdefloep commented:
So most things work, but there is now the problem that an output variable will show up with two dimensions in the output:
this example:
```python
@xs.process
class Foo:
    a = xs.variable(intent="out", dims=(xs.MAIN_CLOCK))

    @xs.runtime(args="main_clock_values")
    def initialize(self, clock_values):
        self.a = clock_values

    @xs.runtime(args="step")
    def run_step(self, step):
        self.a[step] += 1


model = xs.Model({"foo": Foo})
ds_in = xs.create_setup(
    model=model,
    clocks={"clock": range(5)},
    input_vars={},
    output_vars={"foo__a": "clock"},
)
print(ds_in.xsimlab.run(model=model).foo__a)
```
returns:
```python
<xarray.DataArray 'foo__a' (clock: 5)>
array([[1., 1., 2., 3., 4.],
       [1., 2., 2., 3., 4.],
       [1., 2., 3., 3., 4.],
       [1., 2., 3., 4., 4.],
       [1., 2., 3., 4., 4.]])
Coordinates:
  * clock    (clock) int64 0 1 2 3 4
```

benbovy commented on xsimlab/utils.py:
```diff
@@ -15,34 +15,21 @@
+class _MainClockDim:
     """singleton class to be used as main clock dimension: update on runtime
     it has all behaviour that dimensions in a `xr.DataArray` normally have.
     """
```
```suggestion
    """Singleton class to be used as a placeholder of the main clock
    dimension.
    
    It will be replaced by the actual dimension label set during simulation setup
    (i.e., ``main_clock`` argument).

    """
```

benbovy commented on xsimlab/utils.py:
```diff
@@ -15,34 +15,21 @@
-    # so it shows up in print()
-    def __str__(self):
-        return self.main_clock_dim
+        return "MAIN_CLOCK (uninitialized)"
```
```suggestion
        return "MAIN_CLOCK (undefined)"
```

benbovy commented on xsimlab/variable.py:
```diff
@@ -57,7 +57,10 @@ def _as_dim_tuple(dims):
     """
-    if not len(dims):
+    # MAIN_CLOCK is sentinel and does not have length (or zero), so check explicitly
+    if dims == MAIN_CLOCK:
```
```suggestion
    if dims is MAIN_CLOCK:
```

benbovy reviewed (commented):
Nice!!

In your last example, you definitely want to save only one snapshot for `foo__a` instead of saving snapshots for this unchanged value at each time step, i.e., 

```python
ds_in = xs.create_setup(
    ...
    output_vars={"foo__a": None},
)
```

I don't know if there is any use case where it would make sense to update the values a variable with `xs.MAIN_CLOCK` dimension in `run_step` and save multiple snapshots. If there's no case, we could raise when users want to save multiple snapshots, but we can implement this later (let's see how it goes).

Joeperdefloep commented on xsimlab/stores.py:
```diff
@@ -241,6 +241,14 @@ def _create_zarr_dataset(
             )
 
+        # set MAIN_CLOCK placeholder to main_clock dimension
+        if self.mclock_dim in dim_labels and MAIN_CLOCK in dim_labels:
```
This can still happen: when a user declares a variabke with dims 
`(xs.MAIN_CLOCK,'clock')'

benbovy replied:
Ah yes! Hmm then we would probably want to check in `Dataset.xsimlab.update_clocks` that the given clock labels (main clock and other clocks if provided) don't conflict with the dimension labels set in the model. I don't think we do this already.

For convenience we could add a property in `xs.Model` that returns a list of all (unique) dimension labels declared in the model.

Let's fix that in another PR..

Joeperdefloep replied:
not sure if that is the right place though: Am I right that a user may add a process after updating clocks, thus not catching it?

benbovy commented on xsimlab/tests/test_model.py:
```diff
@@ -462,17 +462,30 @@ class Foo:
-            self.a = clock_values
-            assert all(self.a == [0, 1, 2, 3])
+            self.a = clock_values * 2
+            assert all(self.a == [0, 2, 4, 6])
```
nit: numpy testing functions work with any array-like object.

```suggestion
            np.testing.assert_equal(self.a, [0, 2, 4, 6])
```

benbovy commented on xsimlab/tests/test_model.py:
```diff
@@ -462,17 +462,30 @@ class Foo:
             assert all(clock_array[clock_array.dims[0]].data == [0, 1, 2, 3])
 
+        @xs.runtime(args=["step_delta", "step"])
+        def run_step(self, dt, n):
```
You could remove `dt` if you don't use it.

benbovy commented on xsimlab/tests/test_model.py:
```diff
@@ -462,17 +462,30 @@ class Foo:
+
+    # TODO: there is still the problem that the first (0) value of the clock is
+    # set to np.nan in output (still works fine in input) Also, getting
+    # time variables as DataArray as output is not working
```
This should probably be solved when implementing #141, or check [here](https://github.com/benbovy/xarray-simlab/blob/3e8cb81775868e3e7c6495489ba351567e0d7e42/xsimlab/model.py#L774) if the attribute is a DataArray and if it's the case set the cache value with it's `data` property.

Joeperdefloep replied:
what I meant is that all `0` values are set to `nan` in output #172

benbovy commented on xsimlab/tests/test_model.py:
```diff
@@ -486,27 +499,27 @@ def run_step(self):
-        ValueError, match="Do not pass xs.MAIN_CLOCK into input vars dimensions"
-    ):
-        ds_in.xsimlab.run(model=model)
+        with pytest.raises(
+            ValueError, match="Do not pass xs.MAIN_CLOCK into input vars dimensions"
+        ):
+            a = xs.variable(intent="in", dims=xs.MAIN_CLOCK)
+
+        with pytest.raises(
+            ValueError, match="Do not pass xs.MAIN_CLOCK into input vars dimensions"
+        ):
+            b = xs.variable(intent="in", dims=(xs.MAIN_CLOCK,))
+        with pytest.raises(
+            ValueError, match="Do not pass xs.MAIN_CLOCK into input vars dimensions"
+        ):
+            c = xs.variable(intent="in", dims=["a", ("a", xs.MAIN_CLOCK)])
```
nit: this could be tested without being inside a process class.

benbovy reviewed (commented):
Just left some minor comments.

Joeperdefloep commented:
Now, when we print the model, the `MAIN_CLOCK (undefined)` still shows up in the dimensions. Not sure how to solve that (and it sounds like a minor issue)?

benbovy commented:
> Now, when we print the model, the MAIN_CLOCK (undefined) still shows up in the dimensions.

That's intended behavior. The main clock label is not defined yet and may be different from one simulation to another.

benbovy commented on doc/api.rst:
```diff
@@ -154,7 +154,7 @@ Variable
 .. autosummary::
    :toctree: _api_generated/
-
+   MAIN_CLOCK
```
Hmm, adding `MAIN_CLOCK` here yields an error when building the docs with Sphinx (as a result the whole `Variable` section is empty).

Joeperdefloep replied:
hmm I've checked that, and I don't get an error. That is, I did get a "could not import module MAIN_CLOCK" or something,  when I didn't first `python -m pip install .`, but after that it built properly.
e.g. `make html` worked properly without errors.

benbovy reviewed (commented):
Sorry for the long wait @Joeperdefloep. This looks all good to me! Just one remaining issue with the doc build. Maybe we can remove `MAIN_CLOCK` from `api.rst` here and address this later.

benbovy commented on .vscode/settings.json:
```diff
@@ -0,0 +1,3 @@
+{
```
Could you remove this file, please? It should be ignored in the next PRs (#169).

benbovy reviewed (commented):
Two last things and it should be all good:

- comment below
- could you update `whats_new.rst`, please?

benbovy commented:
Excellent @Joeperdefloep! Many thanks for your great work here, and thanks for your patience!
